### ISSUE TITLE
"Uninstalling old version..." message only appears if the APK didn't change (!)

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#7240]
If I run `rm -rf build; flutter run`, then I get:
```
Launching lib/main.dart on Nexus 6 in debug mode...
Building APK in debug mode (android-arm)...         5058ms
Installing build/app.apk...                         7213ms
```
...and then the app launches.

If I run `flutter run` immediately again, I get:

```
Launching lib/main.dart on Nexus 6 in debug mode...
Uninstalling old version...
Installing build/app.apk...                         4919ms
```
...and then the app launches.

That makes little sense. We (a) shouldn't need to upload a new APK if the old one is identical, and (b) if we're installing an APK, we should always be uninstalling it, why would we only uninstall it in some cases?

**Comment by dantrevino** [comment]
The old version is not getting uninstalled.  So when I re-build, my code updates are not being reflected.

**Comment by Hixie** [comment]
Moving out of m3 since this doesn't fit our current priorities.

**Comment by zanderso** [comment]
@jonahwilliams has done work in this area relatively recently (build system). Is this stale?

**Comment by jonahwilliams** [comment]
Let me see if I can verify this happening

**Comment by jonahwilliams** [comment]
I can't reproduce this anymore, it looks like we're correctly not updating the APK now. Yay


```
Running "flutter pub get" in hello_world...                         0.5s
Launching lib/main.dart on Pixel 4 in debug mode...
Running Gradle task 'assembleDebug'...                                  
Running Gradle task 'assembleDebug'... Done                        23.8s
✓ Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...                         5.2s
Syncing files to device Pixel 4...                                  95ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on Pixel 4 is available at: http://127.0.0.1:50516/mSCdjoBMUnQ=/
I/flutter ( 3165): >>> GET /api/v2/text/update_intervals (body length: null)
I/flutter ( 3165): Skipping background update: The DNS server is down or the Twitarr server is non-existent.


Application finished.
jonahwilliams-macbookpro2:hello_world jonahwilliams$ flutter run
Launching lib/main.dart on Pixel 4 in debug mode...
Running Gradle task 'assembleDebug'...                                  
Running Gradle task 'assembleDebug'... Done                         1.6s
✓ Built build/app/outputs/apk/debug/app-debug.apk.
Syncing files to device Pixel 4...                                 111ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on Pixel 4 is available at: http://127.0.0.1:50594/DYhlXucsFTA=/

```
